File: 3-SystemCalls/fibonacci.cpp
// Author: Stéphane Dorotich
// CPSC 457
// Use freely

#include <iostream>

void usage_error(const char * progName) {
    std::cout << "Usage\n\t" << progName << " i\t\t Where i is the ith number of the fibonacci sequence you want to compute.\n";
    exit(-1);
}

// Credit: https://www.javatpoint.com/fibonacci-series-in-cpp
int64_t fib(int i) {
    if (i < 2) return i;
    int64_t n1 = 0, n2 = 1, n3;
    for (int n = 2; n <= i; n++) {
        n3 = n1 + n2;
        n1 = n2;
        n2 = n3;
    }
    return n3;
} 

int main(int argc, const char ** argv) {
    if (argc != 2) usage_error(argv[0]);

    try {
        int i = std::stoi(argv[1]);
        std::cout << fib(i) << std::endl;
    } catch (const std::invalid_argument & ex) {
        usage_error(argv[0]);
    }
}

File: 3-SystemCalls/fibonacci.py
import sys

def usage_error():
    print("Usage:\n\tpython fibonacci.py i\t\twhere i is the ith number of the fibonacci sequence to compute.")
    exit(-1)

def fib(n):
    if (n < 2):
        return n
    n1 = 0
    n2 = 1
    for _ in range(2, n+1):
        n3 = n1 + n2
        n1 = n2
        n2 = n3
    return n3


if __name__ == "__main__":
    if (len(sys.argv) != 2):
        usage_error()
    try:
        n = int(sys.argv[1])
        print(str(fib(n)))
    except:
        usage_error()